APSHTTPClient/APSHTTPRequest.m

```diff
@@ -369,8 +411,49 @@ -(void)URLSession:(nonnull NSURLSession *)session didReceiveChallenge:(nonnull N
 {
     DebugLog(@"%s", __PRETTY_FUNCTION__);
     
+    BOOL useSubDelegate = (self.connectionDelegate != nil && [self.connectionDelegate respondsToSelector:@selector(URLSession:task:didReceiveChallenge:completionHandler:)]);
+    
+    if(useSubDelegate && [self.connectionDelegate respondsToSelector:@selector(willHandleChallenge:forSession:)]) {
+        useSubDelegate = [self.connectionDelegate willHandleChallenge:challenge forSession:session];
+    }
+    
+    if(useSubDelegate) {
+        @try {
+            [self.connectionDelegate URLSession:session task:self.task didReceiveChallenge:challenge completionHandler:completionHandler];
+        }
+        @catch (NSException *exception) {
+            if (self.task != nil) {
+                [self.task cancel];
+                
+                NSMutableDictionary *dictionary = nil;
+                if (exception.userInfo) {
+                    dictionary = [NSMutableDictionary dictionaryWithDictionary:exception.userInfo];
+                } else {
+                    dictionary = [NSMutableDictionary dictionary];
+                }
+                if (exception.reason != nil) {
+                    [dictionary setObject:exception.reason forKey:NSLocalizedDescriptionKey];
+                }
+                
+                NSError* error = [NSError errorWithDomain:@"APSHTTPErrorDomain"
+                                                     code:APSRequestErrorConnectionDelegateFailed
+                                                 userInfo:dictionary];
+                
+                
+                
+                [self URLSession:self.session didBecomeInvalidWithError:error];
+                completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, nil);
+            }
+        }
+        @finally {
+            //Do nothing
+        }
+        return;
+    }
+    
     if (challenge.previousFailureCount) {
         [challenge.sender cancelAuthenticationChallenge:challenge];
+        completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, nil);
```

Comment:
This end of in handling the auth-challenge multiple times. See the below lines.

Reply:
Similar implementation was there in NSURConnection too. I think it will not create any problem as completionHandler is passing NSURLSessionAuthChallengeCancelAuthenticationChallenge, which will cancel the authentication challenge.